File: languages/typescript/extensions/content/sdk.ts
/**
 * CONTENT — the SDK face.
 *
 * **There is no conformance gate on this face, anywhere, and per the operator that
 * is correct rather than a defect: the SDK is a convention.** So the extension is
 * its own instrument — the handler's conformance path runs THROUGH these functions
 * rather than beside them, and `EXTENSION.toml [sdk].*.reached_by` records, per
 * operation, which check reaches it. An operation with an empty `reached_by` is
 * genuinely unmeasured; that set is meant to stay near-empty, and today it holds
 * `AtPeer` (the oracle never exercises namespace scoping) and
 * `reassembleUnderCapability`.
 */

import {
  Ecf,
  Entity,
  type ContentStore,
  type EntityTree,
  type HandlerContext,
} from "entity-core-protocol-typescript";

import { reassembleContent, type ReassembleResult } from "./internal/reassemble.js";
import { ContentTypes, CONTENT_PATTERN } from "./types.js";

/** Lowercase hex INCLUDING the leading format byte (§6.4.2 / core §3.5). */
export function hashHexWithFormat(hash: Uint8Array): string {
  return Array.from(hash, (b) => b.toString(16).padStart(2, "0")).join("");
}

// ── EnsureClosure — §3.3 verify_content ─────────────────────────────────────────

export type ClosureVerdict =
  | { readonly complete: true; readonly totalSize: number; readonly chunkCount: number }
  | { readonly complete: false; readonly code: "missing_chunk" | "empty_chunk"; readonly hash: Uint8Array }
  | { readonly complete: false; readonly code: "size_mismatch" | "blob_not_found" | "not_a_blob"; readonly hash: Uint8Array };

/**
 * §3.3 — completeness and total-size consistency over a blob already in the store.
 *
 * §3.7 classifies this **Conformance**: "implementations MUST agree on whether a
 * blob is complete ... the completeness verdict is cross-peer-uniform." So the
 * three failure modes are named exactly as §3.3 names them, and the order matters:
 * a blob whose first chunk is missing reports `missing_chunk`, not `size_mismatch`,
 * even though the totals also disagree.
 *
 * Per-chunk size is deliberately NOT validated against `chunk_size` — §3.3 says so
 * outright, because content-defined chunking produces variable-size chunks and the
 * entity hash already guarantees chunk integrity.
 */
export function ensureClosure(store: ContentStore, blobHash: Uint8Array): ClosureVerdict {
  const blob = store.get(blobHash);
  if (blob === undefined) return { complete: false, code: "blob_not_found", hash: blobHash };
  if (blob.type !== ContentTypes.Blob) return { complete: false, code: "not_a_blob", hash: blobHash };

  const chunkHashes = Ecf.asArray(Ecf.require(blob.data, "chunks")).map((h) => Ecf.asBytes(h));
  let total = 0;
  for (const chunkHash of chunkHashes) {
    const chunk = store.get(chunkHash);
    if (chunk === undefined) return { complete: false, code: "missing_chunk", hash: chunkHash };
    const payload = Ecf.asBytes(Ecf.require(chunk.data, "payload"));
    if (payload.length === 0) return { complete: false, code: "empty_chunk", hash: chunkHash };
    total += payload.length;
  }

  const declared = Number(Ecf.asUint(Ecf.require(blob.data, "total_size")));
  if (total !== declared) return { complete: false, code: "size_mismatch", hash: blobHash };
  return { complete: true, totalSize: total, chunkCount: chunkHashes.length };
}

// ── AtPeer — §6.4.2 Hash Tree Presence ──────────────────────────────────────────

/**
 * The §6.4.2 canonical namespace probe: is hash `H` bound at
 * `{namespace}/{hex(H)}` in this peer's tree?
 *
 * `hex(H)` carries the format byte (66 chars under ECFv1-SHA-256, 98 under
 * SHA-384). Dropping it — using the bare 64-char digest — is the mistake §6.4.2
 * calls out by name; it destroys the algorithm discriminator and breaks URL ⇄
 * binding parity with NETWORK §6.5.6. Its length is implied by the leading byte
 * and is never assumed, so this does no length check at all.
 *
 * **Unmeasured by any gate.** The oracle's content category has no namespace
 * check. Stated, not hidden.
 */
export function atPeer(
  tree: EntityTree,
  localPeerId: string,
  namespace: string,
  hash: Uint8Array,
): Entity | undefined {
  return tree.get("/" + localPeerId + "/" + namespace + "/" + hashHexWithFormat(hash));
}

/** The §6.4.2 ingest-side binding. Paired with {@link atPeer}; one convention, two directions. */
export function bindAtPeer(
  tree: EntityTree,
  localPeerId: string,
  namespace: string,
  entity: Entity,
): string {
  const path = "/" + localPeerId + "/" + namespace + "/" + hashHexWithFormat(entity.contentHash);
  tree.put(path, entity);
  return path;
}

// ── Descriptor construction — §2.4 presence rule ────────────────────────────────

export interface DescriptorFields {
  readonly content: Uint8Array;
  readonly mediaType?: string;
  readonly typeRef?: Uint8Array;
  readonly name?: string;
}

/**
 * §2.4 — build a descriptor, enforcing the presence rule the type system cannot
 * express: **at least one of `media_type` or `type_ref` MUST be present.** Both MAY
 * be. Throws rather than emitting an invalid entity, because a descriptor with
 * neither is a content-addressed statement that says nothing and would be
 * indistinguishable from a corrupt one once bound.
 */
export function createDescriptor(fields: DescriptorFields): Entity {
  if (fields.mediaType === undefined && fields.typeRef === undefined) {
    throw new Error("CONTENT §2.4 presence rule: a descriptor MUST carry media_type or type_ref");
  }
  return Entity.create(
    ContentTypes.Descriptor,
    Ecf.map(
      ["content", Ecf.bytes(fields.content)],
      ["media_type", fields.mediaType === undefined ? null : Ecf.text(fields.mediaType)],
      ["type_ref", fields.typeRef === undefined ? null : Ecf.bytes(fields.typeRef)],
      ["name", fields.name === undefined ? null : Ecf.text(fields.name)],
    ),
  );
}

/** §5.3 — the dual-level descriptor path. `{publisher}/system/content/descriptor/{B_hex}/{D_hex}`. */
export function descriptorPath(publisherPeerId: string, blobHash: Uint8Array, descriptor: Entity): string {
  return (
    "/" +
    publisherPeerId +
    "/" +
    ContentTypes.Descriptor +
    "/" +
    hashHexWithFormat(blobHash) +
    "/" +
    hashHexWithFormat(descriptor.contentHash)
  );
}

/**
 * §5.3 integrity check (**MUST**): a consumer fetching a descriptor at
 * `.../{B_hex}/{D_hex}` MUST verify `descriptor.data.content == B`. Mismatch ⇒
 * reject. Two-level defence — the path embeds `B_hex`, the body carries `hash(B)`,
 * and both must agree.
 */
export function descriptorMatchesAnchor(descriptor: Entity, blobHash: Uint8Array): boolean {
  if (descriptor.type !== ContentTypes.Descriptor) return false;
  const carried = Ecf.optBytes(descriptor.data, "content");
  if (carried === null || carried.length !== blobHash.length) return false;
  return carried.every((b, i) => b === blobHash[i]);
}

// ── Reassembly — the §3.4 capability-checking wrapper ────────────────────────────

/**
 * The **only** public route to materialized blob bytes in this module, and the
 * "explicit capability-checking wrapper" §3.4 requires before reassembly may be
 * reachable from outside the handler body at all.
 *
 * **The check is weaker than this comment used to claim.** It said a `HandlerContext`
 * cannot be manufactured by a consumer. It can: the class is exported with a public
 * constructor, and this repo's own `test/handler.test.ts` builds one. What the wrapper
 * demands is a context naming this pattern and carrying SOME capability — both
 * caller-supplied — and it does not check that capability against the blob or the
 * namespace. Recorded in `EXTENSION.toml [substrate.capability_wrapper]` (corrected
 * 2026-09-12 by the cross-port review); not fixed here, because the fix is the same
 * design question on all three ports.
 *
 * The two assertions below are defence-in-depth against the one way that could be
 * false — a context from a DIFFERENT handler's dispatch being passed in, which
 * would carry a capability scoped to somebody else's pattern.
 */
export function reassembleUnderCapability(ctx: HandlerContext, blobHash: Uint8Array): ReassembleResult {
  if (ctx.pattern !== CONTENT_PATTERN) {
    throw new Error(
      `CONTENT §3.4: reassembly requires a ${CONTENT_PATTERN} handler context; got '${ctx.pattern}'`,
    );
  }
  if (ctx.callerCapability === null) {
    throw new Error("CONTENT §3.4: reassembly requires a cap-checked dispatch; context carries no caller capability");
  }
  return reassembleContent(ctx.peer.contentStore, blobHash);
}

File: languages/rust/extensions/content/tests/sdk.rs
//! §3.3 closure, §2.4 / §5.3 descriptors, §6.4.2 namespace binding — and the positive
//! half of the §3.4 export-surface check.
//!
//! An integration test in `tests/` sees the crate exactly as a third party does: only
//! the public surface. That is why the §3.4 assertion lives here rather than in a unit
//! test beside the code — a `#[cfg(test)] mod tests` inside `src/` would see the
//! private module and prove nothing about the boundary.

use entity_content::{
    create_blob_fixed, create_descriptor, descriptor_matches_anchor, descriptor_path,
    ensure_closure, hash_hex_with_format, store_blob, ClosureVerdict, DESCRIPTOR,
};
use entity_core_protocol::peer::handler::HandlerContext;
use entity_core_protocol::peer::model::Entity;
use entity_core_protocol::peer::store::Store;
use entity_core_protocol::value::{Key, Value};

// ── §3.3 verify_content ──────────────────────────────────────────────────────

#[test]
fn a_complete_blob_is_complete() {
    let store = Store::new();
    let blob = create_blob_fixed(b"abcdefghijklmnop", 4);
    let h = store_blob(&store, &blob);
    match ensure_closure(&store, &h) {
        ClosureVerdict::Complete {
            total_size,
            chunk_count,
        } => {
            assert_eq!(total_size, 16);
            assert_eq!(chunk_count, 4);
        }
        other => panic!("expected Complete, got {other:?}"),
    }
}

#[test]
fn a_missing_chunk_reports_missing_chunk_and_not_size_mismatch() {
    // §3.3's ORDER is normative: a blob whose first chunk is missing reports
    // `missing_chunk`, even though the totals also disagree. §3.7 classifies the
    // completeness verdict as cross-peer-uniform, so the code a peer returns is part
    // of the contract and not a diagnostic detail.
    let store = Store::new();
    let blob = create_blob_fixed(b"abcdefghijklmnop", 4);
    // Store the manifest but NOT the chunks.
    store.put_entity(&blob.blob);
    match ensure_closure(&store, &blob.blob.hash) {
        ClosureVerdict::Incomplete { code, hash } => {
            assert_eq!(code, "missing_chunk");
            assert_eq!(hash, blob.chunks[0].hash);
        }
        other => panic!("expected Incomplete, got {other:?}"),
    }
}

#[test]
fn a_declared_total_that_disagrees_is_size_mismatch() {
    // NEGATIVE CONTROL for the test above: with every chunk present, the SAME function
    // must reach the size check and report the other code. Without this, "it said
    // missing_chunk" is consistent with a function that always says missing_chunk.
    let store = Store::new();
    let real = create_blob_fixed(b"abcdefghijklmnop", 4);
    for c in &real.chunks {
        store.put_entity(c);
    }
    // A manifest naming the same chunks but lying about the total.
    let lying = Entity::make(
        "system/content/blob",
        Value::Map(vec![
            (Key::Text("total_size".into()), Value::UInt(99)),
            (Key::Text("chunk_size".into()), Value::UInt(4)),
            (Key::Text("chunking".into()), Value::UInt(0)),
            (
                Key::Text("chunks".into()),
                Value::Array(
                    real.chunks
                        .iter()
                        .map(|c| Value::Bytes(c.hash.clone()))
                        .collect(),
                ),
            ),
        ]),
    );
    store.put_entity(&lying);
    match ensure_closure(&store, &lying.hash) {
        ClosureVerdict::Incomplete { code, .. } => assert_eq!(code, "size_mismatch"),
        other => panic!("expected size_mismatch, got {other:?}"),
    }
}

#[test]
fn a_non_blob_entity_is_not_a_blob() {
    let store = Store::new();
    let e = Entity::make("system/type", Value::Map(vec![]));
    store.put_entity(&e);
    match ensure_closure(&store, &e.hash) {
        ClosureVerdict::Incomplete { code, .. } => assert_eq!(code, "not_a_blob"),
        other => panic!("expected not_a_blob, got {other:?}"),
    }
}

// ── §6.4.2 / core §3.5 the format byte ───────────────────────────────────────

#[test]
fn hash_hex_includes_the_leading_format_byte() {
    // 66 chars under ECFv1-SHA-256. Dropping the format byte -- the bare 64-char digest
    // -- is the mistake §6.4.2 calls out by name: it destroys the algorithm
    // discriminator and breaks URL-to-binding parity with NETWORK §6.5.6.
    let e = Entity::make("system/type", Value::Map(vec![]));
    let hex = hash_hex_with_format(&e.hash);
    assert_eq!(hex.len(), 66, "expected 33 bytes of hex, got {}", hex.len());
    assert!(
        hex.starts_with("00"),
        "format byte 0 (ecfv1-sha256) must lead"
    );
}

// ── §2.4 / §5.3 descriptors ──────────────────────────────────────────────────

#[test]
fn a_descriptor_with_neither_media_type_nor_type_ref_is_refused() {
    // The §2.4 presence rule the type system cannot express. A descriptor with neither
    // is a content-addressed statement that says nothing, and once bound it is
    // indistinguishable from a corrupt one.
    assert!(create_descriptor(&[0u8; 33], None, None, Some("x")).is_err());
    assert!(create_descriptor(&[0u8; 33], Some("text/plain"), None, None).is_ok());
    assert!(create_descriptor(&[0u8; 33], None, Some(&[0u8; 33]), None).is_ok());
}

#[test]
fn the_five_three_integrity_check_rejects_a_mismatched_anchor() {
    let blob = create_blob_fixed(b"payload", 4);
    let d = create_descriptor(
        &blob.blob.hash,
        Some("application/octet-stream"),
        None,
        None,
    )
    .unwrap();
    assert!(descriptor_matches_anchor(&d, &blob.blob.hash));
    // NEGATIVE CONTROL: the same function must reject a different anchor, or "it
    // matched" says nothing.
    assert!(!descriptor_matches_anchor(&d, &[9u8; 33]));
    // And a non-descriptor entity is rejected on type before content.
    let not_a_descriptor = Entity::make("system/type", Value::Map(vec![]));
    assert!(!descriptor_matches_anchor(
        &not_a_descriptor,
        &blob.blob.hash
    ));
}

#[test]
fn the_descriptor_path_is_dual_level_and_both_levels_carry_the_format_byte() {
    let blob = create_blob_fixed(b"payload", 4);
    let d = create_descriptor(&blob.blob.hash, Some("text/plain"), None, None).unwrap();
    let path = descriptor_path("PEERID", &blob.blob.hash, &d);
    assert_eq!(
        path,
        format!(
            "/PEERID/{DESCRIPTOR}/{}/{}",
            hash_hex_with_format(&blob.blob.hash),
            hash_hex_with_format(&d.hash)
        )
    );
}

// ── §3.4 — the positive half of the export-surface check ─────────────────────

#[test]
fn the_only_public_route_to_bytes_demands_the_dispatchers_context_and_a_target() {
    // POSITIVE CONTROL for `src/bin/deep_import.rs`, which must FAIL to compile. If
    // this file did not compile, that failure would be indistinguishable from the
    // fixture's and the §3.4 claim would rest on a broken build (D15).
    //
    // RE-ANCHORED 2026-09-16. This assertion used to name `&DispatchAuthority` — a
    // crate-private token that was unforgeable and proved the wrong proposition: *you
    // came through this crate's handler*, whose `handle_op` is also callable in-process.
    // It now names `&HandlerContext`, which only the peer's dispatcher can construct
    // (`Peer::route`, the one construction site in the peer tree) and only after §5.2's
    // `check_permission` ALLOWs.
    //
    // **The `&str` in the middle is the half that was missing entirely.** §3.4 routes
    // materialization through cap-scoped surfaces — namespace-scoped `system/content:get`
    // or tree-path-scoped `local/files:read` — and both are scoped to a PATH. The old
    // signature had nowhere to put one, so the wrapper could not have checked a
    // capability against anything even if it had tried to.
    //
    // **And `&Store` is GONE from the signature, which is a strengthening rather than a
    // tidy-up.** A caller who could pass any store could hand this function one the
    // dispatcher never authorized anything against, making the check — had one existed —
    // a check about the wrong subject. The store now comes from the context.
    //
    // The assertion is on the SIGNATURE, which is what the boundary is made of. Naming
    // the function without calling it is the whole test: it compiles only if the public
    // surface is what we claim. The DECISION behind it is measured in both directions in
    // `src/sdk.rs`'s `authorization_tests`, which is a unit module because this file —
    // being a third party, correctly — cannot construct a context to drive it with.
    let f: fn(&HandlerContext<'_>, &str, &[u8]) -> Result<Vec<u8>, (&'static str, Vec<u8>)> =
        entity_content::reassemble_under_capability;
    let _ = f;
}
